[export] add non-strict training IR

pianpwk opened the pull request:
Summary: Adds non-strict implementation of training IR export. Any expected non-strict training IR failures are also either existing strict training IR or non-strict failures (no new failures added). 4 strict training IR failures also resolved.

Refraining from unifying export/export_for_training, per @ydwu4's feedback :)

Test Plan: added test_export_training_ir_to_run_decomp_non_strict.py for non-strict training IR

Differential Revision: D59349454

tugsbayasgalan commented on torch/export/_trace.py:
```diff
@@ -1697,7 +1733,7 @@ def _export_for_training(
         # nn_module_stack
         if node.op not in ["placeholder", "output"]:
-            for key, (fqn, mod_cls) in node.meta["nn_module_stack"].items():
+            for key, (fqn, mod_cls) in node.meta.get("nn_module_stack", {}).items():
```
Hmm i wonder what function guarantees that nn_module_stack is populated? We should probably fix it there?

pianpwk replied:
Ah, deleted because it seems like we don't actually need this.

tugsbayasgalan commented on torch/export/_trace.py:
```diff
@@ -1550,13 +1573,15 @@ def forward(self, *args, **kwargs):
-            )
-            sig.buffers_to_mutate = pytree.tree_map(_strip_root, sig.buffers_to_mutate)
+            if sig:
+                sig.parameters = pytree.tree_map(_strip_root, sig.parameters)
```
Hmm i feel like you need to this for export_for_training as well? We do this so that we can unwrap the top level module (because make_fx only accepts tuple output).

pianpwk replied:
I guess we don't produce a AOTAutograd graph signature in the training path yet. Gonna tag you in another comment.

tugsbayasgalan reviewed (commented):
Awesome job! left some comments.

pianpwk commented on torch/export/_trace.py:
```diff
@@ -678,53 +721,98 @@ def make_argument_spec(i, node) -> ArgumentSpec:
 
-    is_joint = graph_signature.backward_signature is not None
+    # TODO(pianpwk): maybe rely on AOTAutograd for unifying this? Very ugly right now
+    if is_training:
```
@tugsbayasgalan do you know if we can rely on AOTAutograd (e.g. run_functionalized_fw_and_collect_metadata) to get this signature? It seems like we might(?) need to rely on it to handle some errors in mutation + aliasing too. Not sure if having this make_fx() call within this file is the right thing to do.

pianpwk commented on torch/export/exported_program.py:
```diff
@@ -379,6 +379,7 @@ def _decompose_and_get_gm_with_new_signature_constants(
             fake_args_unwrapped[1],
             fake_params_buffers,
             constant_attrs,
+            is_training=False,
```
maybe this flag should be named something else

pianpwk commented on torch/export/_trace.py:
```diff
@@ -602,6 +605,46 @@ def _compiling_state_context():
+        
+        flat_fn, out_spec = create_tree_flattened_fn(functional_call, params_buffers_args, kwargs)
+        flat_args, in_spec = pytree.tree_flatten((params_buffers_args, kwargs))
+        wrapped_fn = lambda *args: tuple(flat_fn(*args))
```
basically @functools.wraps decorator, without creating an inner function

pianpwk commented on torch/export/_trace.py:
```diff
@@ -1803,6 +1513,8 @@ def _export(
         preserve_module_call_signature: A list of submodule paths for which the original
             calling conventions are preserved as metadata.
 
+        is_training: ...
```
I don't know enough about training IR to fill this in

pianpwk commented on torch/export/_trace.py:
```diff
@@ -683,53 +736,59 @@ def make_argument_spec(i, node) -> ArgumentSpec:
         ],
-        input_tokens=graph_signature.input_tokens,
-        output_tokens=graph_signature.output_tokens,
+        # TODO(pianpwk): tokens/custom objects aren't well handled for training IR
```
maybe good idea to patch test_torchbind.py for this

pianpwk commented on test/export/test_export_training_ir_to_run_decomp_non_strict.py:
```diff
@@ -0,0 +1,52 @@
+# Owner(s): ["oncall: export"]
```
Maybe this should just be merged into the strict training IR file? That would run both strict/non-strict at once

ydwu4 replied:
sounds good to me

pianpwk commented on torch/export/_trace.py:
```diff
@@ -602,6 +605,46 @@ def _compiling_state_context():
             torch.compiler._is_compiling_flag = old_value
 
+    def _make_fx_helper(mod, args, kwargs, **flags):
+        # TODO(pianpwk): have a better story for this once _export_for_training comes together
```
or at least a name

pianpwk commented on torch/export/_trace.py:
```diff
@@ -683,53 +736,59 @@ def make_argument_spec(i, node) -> ArgumentSpec:
+    num_tokens = len(input_names) - params_len - len(flat_fake_args)
+    total_non_user_inputs = params_len + num_tokens
+    is_joint = not is_training and graph_signature.backward_signature is not None
 
```
pretty ugly

ydwu4 commented on torch/export/_trace.py:
```diff
@@ -1684,17 +1735,6 @@ def _export_for_training(
     fake_mode = export_artifact.fake_mode
     module_call_specs = export_artifact.module_call_specs
 
-    # TODO(tmanlaibaatar) Not sure why i need this, but need to re-normalize it.
```
What happened here?

pianpwk replied:
Ah, calling make_fx(record_module_stack=True) goes through ModuleStackTracer, which will do this when the node is created

ydwu4 commented on torch/export/_trace.py:
```diff
@@ -1541,13 +1575,15 @@ def forward(self, *args, **kwargs):
-                _strip_root, sig.inputs_to_parameters
-            )
-            sig.buffers_to_mutate = pytree.tree_map(_strip_root, sig.buffers_to_mutate)
+            if sig is not None:
```
Can you assert that only in training path sig can be None and add a TODO probably?

ydwu4 reviewed (approved):
Overall looks good. Left some comments.

zhxchen17 commented:
I feel we definitely to clean up the implementation a bit... but overall this change looks ok to me and we should revisit the complexity of torch.export.export some time later

pianpwk commented:
@pytorchbot merge

pianpwk committed:
[export] add non-strict training IR (#130062)

Summary:
Adds non-strict implementation of training IR export. Any expected non-strict training IR failures are also either existing strict training IR or non-strict failures (no new failures added). 4 strict training IR failures also resolved.

Refraining from unifying export/export_for_training, per ydwu4's feedback :)

Pull Request resolved: https://github.com/pytorch/pytorch/pull/130062

Test Plan: added test_export_training_ir_to_run_decomp_non_strict.py for non-strict training IR

Reviewed By: ydwu4

Differential Revision: D59349454

Pulled By: pianpwk
